setup.cfg

```diff
@@ -19,3 +19,39 @@ exclude = jedi/third_party/* .tox/*
 
 [pycodestyle]
 max-line-length = 100
+
+
+[mypy]
+# Ensure generics are explicit about what they are (e.g: `List[str]` rather than
+# just `List`)
+disallow_any_generics = True
+
+disallow_subclassing_any = True
+
+# Ensure that optional types are explicit
+no_implicit_optional = True
+strict_optional = True
```

Comment:
Setting these like this also gives consumers of Jedi's types the choice to also do this. Unfortunately if we don't do this and have signatures like:
```python
def foo(bar: str = None): ...
```
Then consumer code which wants to have `strict_optional = True` enabled will end up needing to work around our signatures being like this.

IMO being explicit, like:
```python
def foo(bar: Optional[str] = None): ...
```
is clearer anyway.

Reply:
For parso I also activated mypy now (in https://github.com/davidhalter/parso/pull/146). I feel like I like mypy's defaults here. `strict_optional = True` is probably a very good idea, but I feel like `no_implicit_optional = False` saved me a lot of repetition (and space in function signatures!). So I would prefer that.

Reply:
I definitely used to prefer `no_implicit_optional = True` only for the clarity it brought (there only being one way to type something as being optional and that being explicitly separate from whether or not it had a default value), though I now regard it as almost a flaw in `mypy` that the option exists.

I've tried a number of different ways to work around this in a client, but I'm not even sure it's possible at the moment. I've raised https://github.com/python/mypy/issues/9208 to see if there's a recommended approach.

For now though, at least for the public API, I think it's pretty much necessary to be explicit about the `Optional`-ness of parameters.
Would you be happy with having this just for the public API? (though I realise on asking that that I don't actually know where the public API boundaries are)

Reply:
I understand that it can be annoying that mypy doesn't support this case properly. However, I feel like that's mypy's problem, not ours. Upstream projects can still set `no_implicit_optional = False` if they don't want to deal with some of mypy's issues.

Reply:
Just went through this again, this is probably the last open issue IMO.

Reply:
So there were two outcomes from https://github.com/python/mypy/issues/9208:
- turns out the issue I was having was indeed a bug (now fixed in `master`), but
- the `mypy` team plan to change the default here to `no_implicit_optional = False` and deprecate the current default

As the underlying issue is now fixed, I'm less worried about creating breakage for consumers of the library.

cce3ecb1e45fe0c9ab0ff4d1e1e08de501c6ec56 drops these configs allowing us to go with the defaults.